src/Interpreters/CollectJoinOnKeysVisitor.cpp

```diff
@@ -78,22 +78,56 @@ void CollectJoinOnKeysMatcher::visit(const ASTFunction & func, const ASTPtr & as
     {
         ASTPtr left = func.arguments->children.at(0);
         ASTPtr right = func.arguments->children.at(1);
-        auto table_numbers = getTableNumbers(ast, left, right, data);
-        data.addJoinKeys(left, right, table_numbers);
+        auto table_numbers = getTableNumbers(left, right, data);
+        if (table_numbers.first != 0)
+        {
+            data.addJoinKeys(left, right, table_numbers);
+            if (!data.new_on_expression)
+                data.new_on_expression = ast->clone();
+            else
+                data.new_on_expression = makeASTFunction("and", data.new_on_expression, ast->clone());
+        }
+        else
+        {
+            if (!data.new_where_conditions)
+                data.new_where_conditions = ast->clone();
+            else
+                data.new_where_conditions = makeASTFunction("and", data.new_where_conditions, ast->clone());
+
+            data.move_to_where = true;
+        }
+
     }
     else if (inequality != ASOF::Inequality::None)
     {
         if (!data.is_asof)
-            throw Exception("JOIN ON inequalities are not supported. Unexpected '" + queryToString(ast) + "'",
-                            ErrorCodes::NOT_IMPLEMENTED);
+        {
+            ASTPtr left = func.arguments->children.at(0);
+            ASTPtr right = func.arguments->children.at(1);
+            auto table_numbers = getTableNumbers(left, right, data);
+            if (table_numbers.first != 0)
```

Comment:
Maybe it is better to make this case more explicit. E.g. return pair with `table_numbers.first == table_numbers.second` and check this condition or add some special return value for it.
As I understand `0, 0` can be returned from `getTableNumbers` (but will be rejected later). Can we ensure that we don't we mix some cases? I'm trying to find counterexample, but it works correctly, but handling that case explicitly may be more readable at least.

Reply:
Let's only change check to `table_numbers.first == table_numbers.second` without passing pointer to bool

Reply:
ok, I have change it in the new commit.
And I think getTableForIdentifiers() should not return 0, because if there is a table that is not related to the joinedTables, it will be checked in TreeRewrite.cpp:appendSelect function before collectJoinOnKeysVisitors. So I add an `assert(table_number != 0)` in getTableForIdentifiers().

Reply:
Assert shouldn't check this condition because it means server fail (it actually fails server in debug), but exception only show error to user. So, if it's checked in `appendSelect` and throws an exception it is enough. 
Example of such situation found by fuzzer, see `AST fuzzer (debug) — Assertion `table_number != 0' failed` https://clickhouse-test-reports.s3.yandex.net/19685/e1359b01a1cc34c7a6e5fead6568b6ecae5ba0a9/fuzzer_debug/report.html#fail1

Reply:
fuzzer is an amazing tool. How can I run fuzzer tests on my server?

And I found the sql ` SELECT * FROM table1 INNER JOIN table2 ON (table1.a = table2.a) AND (table1.a AND (1048575 < 65536) AND 2 AND (1048575 < -1)) AND (table2.b < toUInt32(table1, 10)) WHERE table1.b < 1` that table_number != 0' failed. 
But I try it, it make another error:
```
dell123 :)  SELECT * FROM table1 INNER JOIN table2 ON (table1.a = table2.a) AND (table1.a AND (1048575 < 65536) AND 2 AND (1048575 < -1)) AND (table2.b < toUInt32(table1, 10)) WHERE table1.b <1

SELECT *
FROM table1
INNER JOIN table2 ON (table1.a = table2.a) AND (table1.a AND (1048575 < 65536) AND 2 AND (1048575 < -1)) AND (table2.b < toUInt32(table1, 10))
WHERE table1.b < 1

Query id: e8d1f0c5-01de-4ae5-81e0-c539af9d34cf


0 rows in set. Elapsed: 0.019 sec.

Received exception from server (version 21.3.1):
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: JOIN ON inequalities are not supported. Unexpected 'table2.b < toUInt32(table1, 10)': While processing table2.b < toUInt32(table1, 10).

dell123 :)
```

Reply:
> it make another error

Is it `release` build?

`assert` works only in `debug` build. Also we have similar behavior with `LOGICAL_ERROR` that fail server in debug, but in release only throws error to user.

> How can I run fuzzer tests on my server?

Queries from particular run you can get from logs and run locally as you do it now.

Generating fuzzed queries you can run locally for any base query with passing option `--query-fuzzer-runs=n` to `clickhouse-client`

https://github.com/ClickHouse/ClickHouse/blob/f7dbdc623cc4122f7b03f621e2c3f4c4d745b74f/programs/client/Client.cpp#L2482

Reply:
@vdimir yes, it is release mode. and I build a debug mode, assert works.

Thank you for your guidance^^